dj/api/nodes.py

```diff
@@ -418,85 +410,150 @@ def create_cube_node_revision(
-                for column_name, column_data in data.columns.items()
-            ],
-            parents=[],
-        )
-    elif data.type == NodeType.CUBE:
-        node_revision = create_cube_node_revision(session=session, data=data)
-    else:
-        node_revision = create_node_revision(data, session)
-
-    # Point the node to the new node revision.
     node_revision.node = node
     node_revision.version = (
         str(DEFAULT_DRAFT_VERSION)
-        if data.mode == NodeMode.DRAFT
+        if node_mode == NodeMode.DRAFT
         else str(DEFAULT_PUBLISHED_VERSION)
     )
     node.current_version = node_revision.version
-
     node_revision.extra_validation()
 
     session.add(node)
     session.commit()
+
     newly_valid_nodes = resolve_downstream_references(
         session=session,
-        node_revision=node,
+        node_revision=node_revision,
     )
     propagate_valid_status(
         session=session,
         valid_nodes=newly_valid_nodes,
         catalog_id=node.current.catalog_id,  # pylint: disable=no-member
     )
     session.refresh(node.current)
+
+
+@router.post("/nodes/source/", response_model=NodeOutput, status_code=201)
+def create_source_node(
+    data: CreateSourceNode,
+    session: Session = Depends(get_session),
+    query_service_client: QueryServiceClient = Depends(get_query_service_client),
+) -> NodeOutput:
+    """
+    Create a source node. If columns are not provided, the source node's schema
+    will be inferred using the configured query service.
+    """
+    get_node_by_name(
+        session,
+        data.name,
+        raise_if_not_exists=False,
+        raise_if_exists=True,
+    )
+    node = Node(name=data.name, type=NodeType.SOURCE, current_version=0)
+    catalog = get_catalog(session=session, name=data.catalog)
+
+    # When no columns are provided, attempt to find actual table columns
+    # if a query service is set
+    columns = (
+        [
+            Column(
+                name=column_name,
+                type=ColumnType[column_data["type"]],
+            )
+            for column_name, column_data in data.columns.items()
+        ]
+        if data.columns
+        else None
+    )
+    if not columns:
+        if not query_service_client:
+            raise DJException(
+                message="No table columns were provided and no query "
+                "service is configured for table columns inference!",
+            )
+        columns = query_service_client.get_columns_for_table(
+            data.catalog,
+            data.schema_,  # type: ignore
+            data.table,
+        )
+
+    node_revision = NodeRevision(
+        name=data.name,
+        display_name=data.display_name
+        if data.display_name
+        else generate_display_name(data.name),
+        description=data.description,
+        type=NodeType.SOURCE,
+        status=NodeStatus.VALID,
+        catalog_id=catalog.id,
+        schema_=data.schema_,
+        table=data.table,
```

Comment:
@shangyian I think I missed this when we introduced the node revisions. What if I want to create a source node that points to multiple tables, is that no longer possible?

Reply:
Hmm, if they're truly different tables with different schema/table names, I think you're right that it's no longer possible for the system to automatically swap them out without the user explicitly doing that. These are sort of alt-paths down the DAG right? Is there a way we can have these other tables defined on separate nodes and tied together through some other construct?

Reply:
@betodealmeida hmm, that change you're describing wasn't introduced with the node revisions, but rather with the recent changes to move away from the `database` concept and over to `catalog` instead.

Reply:
But agreed that we should discuss this on Wednesday. If a source node has multiple tables, it will also need to manage any schema discrepancies between the nodes.

Reply:
Ah, thanks @shangyian!

> But agreed that we should discuss this on Wednesday. If a source node has multiple tables, it will also need to manage any schema discrepancies between the nodes.

Yeah, we had that in place — the schema of a source node was the superset of the schemas of its representations. One of the source nodes in the original examples had a column that only existed in GSheets (slow), I think it was called `secret_column`. Any query that used that column would hit GSheets, instead of a faster database, since it was required to fulfill the request.

Reply:
+1 on discussing this early on. We can start on Wed but if we need more time we should schedule a separate chat.

One thing to add to the discussion from me would be checking/enforcing the grain (aka PK) in the source nodes that point to multiple tables. I know having separate columns available in different datasets is ok, but I think accounting for different grain would be tough.